### PR TITLE
zuse: add schnorr signatures

### PR DESCRIPTION
**Pull request by mrdomino**
Jets here: #5510.

### CODE REVIEW

**Review by joemfb — COMMENTED** [review]
This should have some unit tests -- there are official vectors here: https://github.com/bitcoin/bips/tree/master/bip-0340.

I'm not familiar with the domain, but would it be reasonable to refactor this into a generic implementation, parameterized on hash functions, lengths, tag constants, &c?

**Comment by mrdomino** [comment]
I’ll add the test vectors; I tested against them manually from dojo but obviously we want them in the repo.

I’ll ask about the generic implementation as well.

**Comment by mrdomino** [comment]
Re generic implementation: Tim says if it’s good enough for Bitcoin to add in a major update, it’s enough of an update for us for now.

(The C side’s [`secp256k1_schnorrsig_sign`](https://github.com/bitcoin-core/secp256k1/blob/master/src/modules/schnorrsig/main_impl.h#L195) only implements the default parameters; other stuff is supported only in a separate `secp256k1_schnorrsig_sign_custom`. Seems reasonable to take a similar approach with the hoon if in future anyone ever wants to parameterize further.)

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon:2240]: Style nit, this indentation should be:

```
=>  |%
    ++  tagged-hash
```

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon:2242]: The levels of swapping here are frustrating. `+sha-256:sha` swaps it's input and output, so this swaps thrice. It's equivalent to `(rev 3 32 (shax tag))`, but idk if it's worth changing. Just frustrating ...

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon:2284]: This is the same as `(rep 8 [m x.pp t ~])`.

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon:2296]: Also `(rep 8 ...)`

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon:2319]: This is `(end 8 sig)`

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon:2327]: Also `(rep 8 ....)`

File: pkg/arvo/tests/sys/zuse/crypto/secp256k1.hoon
Comment (joemfb) [inline pkg/arvo/tests/sys/zuse/crypto/secp256k1.hoon:318]: In a departure from standard, 3-fixed backstep indentation, test-case vase literals are usually both indented:

```
%+  expect-eq
  !>  expected
  !>  actual
```

**Review by joemfb — CHANGES_REQUESTED** [review]
I've noted some style nits and simpler atom ops. I've glanced at the spec, but haven't closely reviewed for edge cases.

File: pkg/arvo/sys/zuse.hoon
Comment (mrdomino) [reply pkg/arvo/sys/zuse.hoon:2242]: Yep. Root cause is that BIP 340 specifies things the wrong endianness so something is gonna be backwards no matter what. I opted for consistency with the `+sha` usage below.

**Review by joemfb — APPROVED** [review]
Changes look good.

Please get another review for spec sanity check, and hold off on merging 'till the jet is sorted out.

File: pkg/arvo/sys/zuse.hoon
Comment (joemfb) [inline pkg/arvo/sys/zuse.hoon]: It's important to be straightforward to verify that assertions/crashes in the jet match the hoon/nock. Can you add a comment to the source in addition to the commit msg that removed this?

**Comment by tacryt-socryp** [comment]
I have manually reviewed every value in every test vector, and determined that they are identical to those found in the BIP.

File: pkg/arvo/sys/zuse.hoon
Comment (tacryt-socryp) [inline pkg/arvo/sys/zuse.hoon:2287]: directly after 2287, insert an assertion that kp must not be 0

File: pkg/arvo/sys/zuse.hoon
Comment (tacryt-socryp) [inline pkg/arvo/sys/zuse.hoon:2250]: Must assert that input x is in the range 0...p-1

**Comment by tacryt-socryp** [comment]
Aside from the comments that I read, I went over the pseudocode of every line alongside the Hoon and verified that it is all represented here, even the pesky "is infinite check" in point addition which is mirrored here: https://github.com/bitcoin/bips/blob/97e02b2223b21753acefa813a4e59dbb6e849e77/bip-0340/reference.py#L45

**Review by tacryt-socryp — APPROVED** [review]

**Comment by mrdomino** [comment]
added assertions and verified that unit tests still pass with jets disabled

**Pull request merged**
